### ISSUE TITLE
Add Stream Data Type and Infrastructure

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#5132]
I will start from the beginning and work my way to the end.

## Stream Data Type

We begin by defining the following data type:

```scala
enum Stream[t: Type, pur: Bool, ef: Eff]{
    case Stream(Unit -> Option[t] & pur \ ef, Region[pur])
}
```

A `Stream` is indexed by the type of its elements and the effect(s) of advancing the stream. 

Since Flix currently has two effects systems we need both ` pur: Bool` and `ef: Eff`. Cumbersome yes. At some point the effect systems will merge and we will only have `enum Stream[t: Type, ef: Eff]`.

As the reader might have guessed by now, a `Stream` is like an iterator. Inside it we find the `next` function: `Unit -> Option[t] & pur \ ef, Region[pur]`. This function returns the next element in the stream as `Some(x)` or `None` if the stream is empty. 

Using `next` with this signature is slower than using two functions: `hasNext` and `next`, but I choose to start here, because it is conceptually similar, and we can change the internals later anyway. 

## Stream Operations

What can we do with streams? We don't need much to get started (but we will probably want a lot of functionality later).

We need a way to filter:

```scala
def filter(f: a -> Bool & pur1 \ ef1, s: Stream[a, pur2, ef2]): Stream[a, pur1 and pur2, {ef1, ef2}] = ???
```

Filtering is lazy. The signature here is a bit horrendous (but it will get worse).

We also want `flatMap` (this is an efficient way to combine Streams):

```scala
def flatMap(f: a -> Stream[b, pur1, ef1] & pur2 \ ef2, s: Stream[a, pur3, ef3]): Stream[b, pur1 and pur2 and pur3, {ef1, ef2, ef3}] = ???
```

The `flatMap` function is (very) tricky to implement, but I believe it can be done. 

## Streamable Type Class

Next, we want a way to create streams from data structures. The idea is to add a new type class:

```scala
pub class Streamable[t: Type -> Type] {
    pub def stream(r: Region[r], x: t[a]): Stream[a, r, { Pure }] \ r
}
```

We can implement `Streamable` for e.g. `List` and `Set`. We cannot implement it for `MutList` because the shape is wrong, but I believe that can be solved by our associated types idea. Thus, let us assume for now, that we can implement `Streamable` for any collection over `t`. Immediately another issue is `Result[t, r]` and `Map[k, v]`. Here I believe if we swap arguments to `Result` then it can work too and for `Map[k, v]` we can implement it over **the values** of the `Map`. (If you want both keys and values we might need a special function you can call, e.g. `Map.toKeyValueStream` or whatever. So, the hope is that `Stream` can be directly or indirectly implemented for everything which is "collection-like".

## FromStream Type Class

Next, we introduce the `FromStream` (`Collectable`?) type class:

```
class FromStream[m: Type -> Type] {
    pub def collect(s: Stream[a, pur, ef]): m[a] & pur \ ef
}
```

This is a cool one. It says I can turn any `Stream[t, ...]` into any `m[t]` where there is an instance for `m`.

For example
- We can implement `FromStream` on `List` and `Set`. 
- I am not sure about `Result` and `Map` because of the two type parameters.
- I am also not sure about `MutList` (because it needs a region). 
- Probably cannot implement it on `Nel` since some streams are empty.

In any case, any data type that has the shape `m[t]` should work (e.g. List, Set, Chain, maybe Option).

## Putting it together

With all of this machinery, we can now write the following program:

```scala
let c1 = Some("a");
let c2 = List#{"b"};
let c3 = Set#{1, 2};
region rh {
  let s1 = Streamable.stream(rh, c1);
  let s2 = Streamable.stream(rh, c2);
  let s3 = Streamable.stream(rh, c3);
  let pipeline = 
    Stream.flatMap(x -> 
      Stream.flatMap(y -> 
        Stream.flatMap(z -> (x, y, z), s1), s2), s3);
  let result = FromStream.collect(rh, pipeline);
  Set.size(result)
}
```

This program computes the set: `Set#{("a", "b", 1), ("a", "b", 2)}`.

What is important here is that in the last line we call `Set.size` hence type inference will conclude that `FromStream.collect` has to return a `Set`. 

This is great!! But a bit verbose.

## Adding sugar

Let us replace those ugly `flatMaps` by use of a foreach-like construct:

```scala
let c1 = Some("a");
let c2 = List#{"b"};
let c3 = Set#{1, 2};
region rh {
  let s1 = Streamable.stream(rh, c1);
  let s2 = Streamable.stream(rh, c2);
  let s3 = Streamable.stream(rh, c3);
  let pipeline = 
    foreach (x <- c1; y <- c2, z <- c3) 
      yield (x, y, z)
  let result = FromStream.collect(rh, pipeline);
  Set.size(result)
}
```

The region is a bit annoying, but what if every `foreach` is simply wrapped in its own region? And at the same time, we make those `Streamable.stream` calls implicit in the `foreach` (using the region we have introduced). Thus we arrive at:

```scala
let c1 = Some("a");
let c2 = List#{"b"};
let c3 = Set#{1, 2};
let pipeline = 
  foreach (x <- c1; y <- c2, z <- c3) 
    yield (x, y, z)
let result = FromStream.collect(rh, pipeline);
Set.size(result)
```

The `collect` call is really on pipeline, so we have:

```scala
let c1 = Some("a");
let c2 = List#{"b"};
let c3 = Set#{1, 2};
let result = 
  FromStream.collect(rh, (foreach (x <- c1; y <- c2, z <- c3) 
    yield (x, y, z))
Set.size(result)
```

But then we might as well make the `collect` implicit too and we arrive at the final program:

```scala
let c1 = Some("a");
let c2 = List#{"b"};
let c3 = Set#{1, 2};
let result = 
  foreach (x <- c1; y <- c2, z <- c3) 
    yield (x, y, z)
Set.size(result)
```

Let us remind ourselves how this works:

- A `foreach-yield` implicitly introduces a region, so we have `region rh { foreach ... }`.
- A `foreach-yield` implicit wraps that in a call to `collect`, so we have `region rh { FromStream.collect(foreach ..., rh) }`.
- Inside the `foreach`, we simply desugar each `x <- c` into something like `flatMap(x -> ..., Streamable.stream(c, rh)` converting each collection into a fresh stream using our current region handle and into a call to `flatMap`.

***We can iterate on any type that implements `Streamable` and (automatically) return any type that implements `FromStreamable`***

- Ultimately these constructs would replace `Iterator` (i.e. we can remove Iterator and optionally rename Stream to Iterator).
- The `foreach-yield` construct will be different from the `for` construct which is still monadic bind. (We can shuffle the names around if desired.)
- Note that the implementation uses regions and mutable data (which makes it fast), but the overall construct remains pure once we leave the region. In other words, you can use `foreach-yield` in a pure way, if you want. You can, of course, also have side-effects. 

Actionable items:

- [x] Add `Stream` data type with `filter`, `map`, and `flatMap`.
- [x] Add `Streamable` type class and common instances.
- [x] Add `Collectable` (Collectible?) type class and common instances.
- [ ] Add `foreach-yield` syntax to frontend of the compiler and desugar in the `Weeder`.

Later:
- [ ] Add instances for mutable collections (requires associated types).
- [x] Figure out how to deal with `Result[t, e]` / `Map[k, v]`.
- [ ] Figure out how to collect into a mutable collection.

Each PR should come with its own tests.

**Comment by paulbutcher** [comment]
Can you confirm that my understanding is correct:

1. `foreach`-`yield` on immutable data structures will (appear to) be pure. So this is valid:
    
    ```
    def foo(): Set[(Int32, Int32)] =
        foreach(x <- Set#{1, 2}; y <- Set#{3, 4}) yield (x, y)
    ```
    and would return `Set#{(1, 3), (1, 4), (2, 3), (2, 4)}`?
2. Unlike `region` which magically creates both a term-level and type-level variable `r`, the region variable introduced by `foreach`-`yield` wouldn't be visible or accessible to code?
3. The challenge associated with `Map[k, v]` is different from the challenge associated with `Result[t, e]`:
    1. For `Map` it's really a question of whether we iterate over values or key/value pairs by default (and how we provide the alternative if necessary).
    2. For `Result` it's the same issue as we've had all along (the order of the type arguments).

Assuming that I'm right above, is there any solution to handling `Result` *other* than swapping the order of the type arguments?

**Comment by magnus-madsen** [comment]
> 1. foreach-yield on immutable data structures will (appear to) be pure. So this is valid:

Yes, and yes.

> 2. Unlike region which magically creates both a term-level and type-level variable r, the region variable introduced by foreach-yield wouldn't be visible or accessible to code?

There will be both a term and type level region variable, but they will not be accessible, yes.

> 3. The challenge associated with Map[k, v] is different from the challenge associated with Result[t, e]:

Yes. The problem is with the signature of an iterator/streamable, its over a `t` not a `(k, v)`. 
For `Result` its the same issue.

> Assuming that I'm right above, is there any solution to handling Result other than swapping the order of the type arguments?

I am looking for it, but if I don't find it, I will surrender :)

**Comment by paulbutcher** [comment]
> There will be both a term and type level region variable, but they will not be accessible, yes.

Good. I still (strongly) think that we need to get rid of the magic within `region`, and it's good to know that we're not creating more magic!

**Comment by magnus-madsen** [comment]
Notice that the solution is not that complicated: To explain it to someone you could say: The foreach-yield introduces a new region, it then allocates an iterator for each collection in your for, combines them (with flatMap), and then turns everything back into a collection using `collect/Collectable/Collectible`?

**Comment by paulbutcher** [comment]
Yes, I like it from that POV. It's very similar to what Clojure and F# do.

**Comment by magnus-madsen** [comment]
To explain why it needs to introduce a new region, one could say: an iterator/stream is a mutable thing so it needs to live in a region. Since we only need a region for the foreach-yield, the compiler can automatically introduce it.

**Comment by magnus-madsen** [comment]
> Yes, I like it from that POV. It's very similar to what Clojure and F# do.

And Rust!

**Comment by paulbutcher** [comment]
> To explain why it needs to introduce a new region

We wouldn't have to explain this at all if the region is inaccessible (which is a good thing)

**Comment by magnus-madsen** [comment]
> > To explain why it needs to introduce a new region
> 
> We wouldn't have to explain this at all if the region is inaccessible (which is a good thing)

Or provide syntax to name it, should you desire? (Thinking of Scala pattern match "at patterns")

**Comment by paulbutcher** [comment]
> Or provide syntax to name it, should you desire?

That would work for me. And perhaps we could use a similar approach to de-magic `region`?

**Comment by magnus-madsen** [comment]
I think this solution has so many upsides its hard to argue against:

- It gives us the sugar we want.
- It desugars to a small set of understandable functions and interfaces.
- Its very efficient (no intermediate data structures).
- It is pure to the outside.
- It can collect into any collection.
- It avoids and does not confusing anything with Monads--- its pure an abstraction for turning collections into other collections.
- It is reasonably explainable. 
- It supports effects in the for-loop.
- It integrates well into the rest of the language.

The only issue really is that iterating through a `Map[k, v]` will only give you the `v`s. That's annoying, but can be fixed by having `Map.toKeyValueStream` or whatever. (Since this is a type class issue, we may also be able to overcome it in the future).

**Comment by magnus-madsen** [comment]
The following hack could work for any 2-element collection.

We introduce:

```
pub class Streamable2[t: Type -> Type -> Type] {
    pub def stream(r: Region[r], x: t[a, b]): Stream[(a, b), r, { Pure }] \ r
}
```

then IF the foreach-fragment is of the exact form: `for (... (x, y) <- c ...)` (i.e. a pair instead of just a variable) we call `Streamable2.stream` on `c` (instead of `Streamable.stream`).

**Comment by jaschdoc** [comment]
Looks nice.

> The following hack could work for any 2-element collection.
> 
> We introduce:
> 
> ```
> pub class Streamable2[t: Type -> Type -> Type] {
>     pub def stream(r: Region[r], x: t[a, b]): Stream[(a, b), r, { Pure }] \ r
> }
> ```
> 
> then IF the foreach-fragment is of the exact form: `for (... (x, y) <- c ...)` (i.e. a pair instead of just a variable) we call `Streamable2.stream` on `c` (instead of `Streamable.stream`).

If we choose to desugar the `foreach-yield` during `Lowering` instead of `Weeder`, wouldn't we then have sufficient type information to always call the correct type class?

Maybe it is possible to introduce a type class that can convert a `Streamable2` to `Streamable` (sort of like `flatten` on a data structure), so we could effectively support a type constructor of any arity simply by "flattening" the stream repeatedly (i.e. introducing the function calls during `Lowering`). I'm not sure how this should be implemented but just a thought I had when reading through it.

**Comment by JonathanStarup** [comment]
I like it :)

* In "Putting it together" shouldn't the last `flatMap` be `map`?
* Besides the mutable/immutable type class issue and the multiparam Result & map issue it feels like a super solution.
* We should implement the Stream classes for steam itself right? Then `Map.toKeyValueStream` would just return a stream and everything works. Of course even just `Map.toList` would work, even though the overhead is annoying.

**Comment by magnus-madsen** [comment]
> If we choose to desugar the foreach-yield during Lowering instead of Weeder, wouldn't we then have sufficient type information to always call the correct type class?

TLDR: No, you will end up in the swamp. Its chicken versus egg essentially.

The flatten idea can work, but only for a fixed depth. And programmers would not expect this.

**Comment by magnus-madsen** [comment]
@JonathanStarup I think you are right that the last operation (inner most) should be a map.

**Comment by mlutze** [comment]
> The only issue is that iterating through a Map[k, v] will only give you the vs. That's annoying, but can be fixed by having Map.toKeyValueStream or whatever. (Since this is a type class issue, we may also be able to overcome that in the future).

If we're already banking on having associated types, then this is easily fixed by adding one more associated type to the class.

```scala
class Streamable[t: Type] {
    associated type Advance: Effect
    associated type Element: Type
    pub def stream(r: Region[r], x: t): Stream[Element, {Advance, r}] \ {Advance, r}
}

instance Streamable[Map[k, v]] {
    associated type Advance = Pure
    associated type Element = (k, v)
    pub def stream(r: Region[r], x: Map[k, v]): Stream[(k, v), {r}] \ {Advance, r}
}

**Comment by magnus-madsen** [comment]
Very nice 👍

**Comment by magnus-madsen** [comment]
I think if one implements `append` then `flatMap` might be easier to implement.

**Comment by magnus-madsen** [comment]
Would anyone be interested in implementing (some) of this?

I think @jaschdoc can do the sugar part, but that would be one of the last steps.

**Comment by jaschdoc** [comment]
If nobody's working on it when I'm done with the array syntax changes and front end error handling then sure but that may be a while

**Comment by magnus-madsen** [comment]
> If nobody's working on it when I'm done with the array syntax changes and front end error handling then sure but that may be a while

I think you are a bit busy :) But you can help with the syntax once we get that far.

**Comment by magnus-madsen** [comment]
@stephentetley I occurs to me that your discussion of indexed monads at https://gist.github.com/stephentetley/7bee863df3c3a2721fabb2e2c54c8426 seems related to this ticket where `Stream` will store its "delayed effects". 

In particular this:

```
def flatMap(f: a -> Stream[b, pur1, ef1] & pur2 \ ef2, s: Stream[a, pur3, ef3]): Stream[b, pur1 and pur2 and pur3, {ef1, ef2, ef3}] = ???
```

seems closely related to `IxMonad`.

**Comment by stephentetley** [comment]
Yes Magnus - I think that definition of `flatMap` would make Stream a graded monad, that is an indexed monad where the index is monoidal (here the monoid is a set of effects).

I've experimented with classes in the project below, but making using them work with monads in CPS seems hard (for stack safety I think CPS monads are the way to go).

https://github.com/stephentetley/graded-lib

**Comment by stephentetley** [comment]
By the way, Java's Collector interface seems a very useful way of working with streams. 

Possibly with it coming from Java it hasn't received much attention in the functional world[*], but it seems to me its a way of getting something quite like fusion - the "fusion" is very much by hand but you do get it on Streams which are by nature "run once". 

[*] Maybe it's quite like the idea of composable "Beautiful Folds" that got quite a bit of attention a few years ago.

http://squing.blogspot.com/2008/11/beautiful-folding.html

https://hackage.haskell.org/package/ZipFold

**Comment by magnus-madsen** [comment]
I think the above task list can be achieved by a careful refactoring of `Iterable` / `Iterator`.

**Comment by magnus-madsen** [comment]
This now works:

```
def main(): Unit \ IO = 
    region rh {
        let c1 = Set#{1, 2, 3};
        let c2 = List#{1, 2, 3};
        let iter1 = Iterable.iterator(rh, c1);
        let iter2 = Iterable.iterator(rh, c2);
        let result: Chain[(Int32, Int32)] = Collectable.collect(
            Iterator.flatMap(x -> Iterator.flatMap(y -> Iterator.singleton(rh, (x, y)), iter2), iter1)
        );
        println(result)
    }
```

Honestly amazing!

**Comment by magnus-madsen** [comment]
@JonathanStarup Can you add the above example to the region paper, I think its very cool :)

**Comment by magnus-madsen** [comment]
@paulbutcher We are very closing having the for-loop-thingy for the chess example! 🎉 💪

**Comment by magnus-madsen** [comment]
Fixed in master.
